apis/mattermost/v1beta1/mattermost_types.go

```diff
@@ -0,0 +1,263 @@
+// https://book.kubebuilder.io/reference/generating-crd.html                  //
+////////////////////////////////////////////////////////////////////////////////
+
+// MattermostSpec defines the desired state of Mattermost
+type MattermostSpec struct {
+	// Size defines the size of the Mattermost. This is typically specified in
+	// number of users. This will override replica and resource requests/limits
+	// appropriately for the provided number of users. This is a write-only
+	// field - its value is erased after setting appropriate values of resources.
+	// Accepted values are: 100users, 1000users, 5000users, 10000users,
+	// and 250000users. If replicas and resource requests/limits are not
+	// specified, and Size is not provided the configuration for 5000users will
+	// be applied. Setting 'Replicas', 'Advanced.Resources', 'Filestore.Replicas',
+	// 'Filestore.Resource', 'Database.Replicas', or 'Database.Resources' will
+	// override the values set by Size. Setting new Size will override previous
+	// values regardless if set by Size or manually.
+	// +optional
+	Size string `json:"size,omitempty"`
+
+	// Image defines the Mattermost Docker image.
+	Image string `json:"image,omitempty"`
+	// Version defines the Mattermost Docker image version.
+	Version string `json:"version,omitempty"`
+	// Replicas defines the number of replicas to use for the Mattermost app
+	// servers.
+	Replicas int32 `json:"replicas,omitempty"`
+	// Optional environment variables to set in the Mattermost application pods.
+	// +optional
+	MattermostEnv []v1.EnvVar `json:"mattermostEnv,omitempty"`
+	// MattermostLicenseSecret is the name of the secret containing a license.
+	// +optional
+	MattermostLicenseSecret string `json:"mattermostLicenseSecret,omitempty"`
+	// IngressName defines the name to be used when creating the ingress rules
+	IngressName string `json:"ingressName"`
+	// +optional
+	IngressAnnotations map[string]string `json:"ingressAnnotations,omitempty"`
+	// +optional
+	UseIngressTLS bool `json:"useIngressTLS,omitempty"`
+	// +optional
+	UseServiceLoadBalancer bool `json:"useServiceLoadBalancer,omitempty"`
+	// +optional
+	ServiceAnnotations map[string]string `json:"serviceAnnotations,omitempty"`
+	// +optional
+	ResourceLabels map[string]string `json:"resourceLabels,omitempty"`
+	// Volumes allows for mounting volumes from various sources into the
+	// Mattermost application containers.
+	// +optional
+	Volumes []*v1.Volume `json:"volumes,omitempty"`
+
+	// Advanced contains additional advanced settings for Mattermost.
+	Advanced Advanced `json:"advanced,omitempty"`
+
+	// External Services
+	Database      Database      `json:"database,omitempty"`
+	Filestore     Filestore     `json:"filestore,omitempty"`
+	ElasticSearch ElasticSearch `json:"elasticSearch,omitempty"`
+}
+
+// Advanced defines the advanced configuration for Mattermost. These settings
+// should be correct for most types of deployments.
+type Advanced struct {
```

Comment:
I am thinking if we could come up with a better name as `Advanced` is not very descriptive.

Reply:
Yeah, we discussed this a bit in the design doc. The comment was supposed to explain the intention of the object, but open to new ideas if we have them.